sdks/python/apache_beam/dataframe/transforms.py

```diff
@@ -0,0 +1,255 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import
+
+import pandas as pd
+
+import apache_beam as beam
+from apache_beam import transforms
+from apache_beam.dataframe import expressions
+from apache_beam.dataframe import frame_base
+from apache_beam.dataframe import frames  # pylint: disable=unused-import
+
+
+class DataframeTransform(transforms.PTransform):
+  """A PTransform for applying function that takes and returns dataframes
+  to one or more PCollections.
+
+  For example, if pcoll is a PCollection of dataframes, one could write::
+
+      pcoll | DataframeTransform(lambda df: df.group_by('key').sum(), proxy=...)
+
+  To pass multiple PCollections, pass a tuple of PCollections wich will be
+  passed to the callable as positional arguments, or a dictionary of
+  PCollections, in which case they will be passed as keyword arguments.
+  """
+  def __init__(self, func, proxy):
+    self._func = func
+    self._proxy = proxy
+
+  def expand(self, input_pcolls):
+    def wrap_as_dict(values):
+      if isinstance(values, dict):
+        return values
+      elif isinstance(values, tuple):
+        return dict(enumerate(values))
+      else:
+        return {None: values}
+
+    # TODO: Infer the proxy from the input schema.
+    def proxy(key):
+      if key is None:
+        return self._proxy
+      else:
+        return self._proxy[key]
+
+    # The input can be a dictionary, tuple, or plain PCollection.
+    # Wrap as a dict for homogeneity.
+    # TODO: Possibly inject batching here.
+    input_dict = wrap_as_dict(input_pcolls)
+    placeholders = {
+        key: frame_base.DeferredFrame.wrap(
+            expressions.PlaceholderExpression(proxy(key)))
+        for key in input_dict.keys()
+    }
+
+    # The calling convention of the user-supplied func varies according to the
+    # type of the input.
+    if isinstance(input_pcolls, dict):
+      result_frames = self._func(**placeholders)
+    elif isinstance(input_pcolls, tuple):
+      result_frames = self._func(
+          *(value for _, value in sorted(placeholders.items())))
+    else:
+      result_frames = self._func(placeholders[None])
+
+    # Likewise the output may be a dict, tuple, or raw (deferred) Dataframe.
+    result_dict = wrap_as_dict(result_frames)
+
+    result_pcolls = {
+        placeholders[key]._expr: pcoll
+        for key, pcoll in input_dict.items()
+    } | 'Eval' >> DataframeExpressionsTransform(
+        {key: df._expr
+         for key, df in result_dict.items()})
+
+    # Convert the result back into a set of PCollections.
+    if isinstance(result_frames, dict):
+      return result_pcolls
+    elif isinstance(result_frames, tuple):
+      return tuple((value for _, value in sorted(result_pcolls.items())))
+    else:
+      return result_pcolls[None]
```

Comment:
Could this be streamlined with calls to `to_dataframe` and `to_pcollection`?

Reply:
Yes. Done.